src/vtpty/SshSession.cpp

```diff
@@ -0,0 +1,1168 @@
+#endif
+            default: return "Unknown error";
+        }
+    }
+} // namespace
+// }}}
+
+// {{{ SshPtySlave
+class SshPtySlave final: public PtySlave
+{
+  public:
+    void close() override {}
+    [[nodiscard]] bool isClosed() const noexcept override { return false; }
+    bool configure() noexcept override { return true; }
+    bool login() override { return true; }
+    int write(std::string_view) noexcept override { return 0; }
+};
+// }}}
+
+#if defined(_WIN32)
+// Why is it, that Windows API is so much more complicated than POSIX? Extrawurst!
+using socket_handle = crispy::native_handle<SOCKET, INVALID_SOCKET>;
+#else
+using socket_handle = crispy::native_handle<int, -1>;
+#endif
+
+struct SshSession::Private
+{
+    LIBSSH2_SESSION* sshSession = nullptr;
+    LIBSSH2_CHANNEL* sshChannel = nullptr;
+    LIBSSH2_AGENT* sshAgent = nullptr;
+    bool wantsWaitForSocket = false;
+
+    socket_handle sshSocket;
+};
+
+SshSession::SshSession(std::string host,
+                       int port,
+                       std::string username,
+                       std::filesystem::path privateKeyFile,
+                       std::filesystem::path publicKeyFile,
+                       std::filesystem::path knownHostsFile,
+                       bool forwardAgent,
+                       Environment env):
+    _host { std::move(host) },
+    _port { port },
+    _username { std::move(username) },
+    _privateKeyFile { std::move(privateKeyFile) },
+    _publicKeyFile { std::move(publicKeyFile) },
+    _knownHostsFile { std::move(knownHostsFile) },
+    _forwardAgent { forwardAgent },
+    _env { std::move(env) },
+    _ptySlave { std::make_unique<SshPtySlave>() },
+    _p { new Private(), [](Private* p) {
+            delete p;
+        } }
+{
+    libssh2_init(0); // TODO: call only once?
+
+#if defined(SSH_SESSION_NB_IO)
+    libssh2_session_set_blocking(_p->sshSession, 0);
+#endif
+
+    std::atexit([]() { libssh2_exit(); });
+
+    _p->sshSession = libssh2_session_init();
+}
+
+SshSession::~SshSession()
+{
+    close();
+
+    if (_p->sshAgent)
+    {
+        libssh2_agent_disconnect(_p->sshAgent);
+        libssh2_agent_free(_p->sshAgent);
+        _p->sshAgent = nullptr;
+    }
+
+    if (_p->sshChannel)
+    {
+        libssh2_channel_send_eof(_p->sshChannel);
+        libssh2_channel_close(_p->sshChannel);
+        _p->sshChannel = nullptr;
+    }
+
+    if (_p->sshSession)
+    {
+        libssh2_session_disconnect(_p->sshSession, "Normal shutdown");
+        libssh2_session_free(_p->sshSession);
+        _p->sshSession = nullptr;
+    }
+
+#if defined(_WIN32)
+    WSACleanup();
+#endif
+}
+
+void SshSession::setState(State nextState)
+{
+    if (_state == nextState)
+        return;
+
+    // sshLog()("({}) SshSession: state change: {} ~> {}\n", threadName(), _state, nextState);
+
+    _state = nextState;
+}
+
+void SshSession::processState()
+{
+    waitForSocket(std::chrono::seconds(8));
```

Comment:
Is there any specification for this number?

Reply:
oh damn it, i think i should be even get rid of that line for now. 

and no, it was just an arbitrary non-configurable timeout to warid on I/O before resuming work.